### ISSUE TITLE
[Bug] The same version of joern behaves differently on different machines.

### ISSUE THREAD

**Issue by wildoranges** [joernio/joern#4611]
**Describe the bug**
I‘m now using joern 2.0.385. I want to generate cpg.bin for the following C code:
```c
#include <stdio.h>

struct data {
    int flag;
};

static struct data data;

int f(int x)
{
    struct data *p = x ? &data : NULL;
    return (p && p->flag);
}

static void acl_read_cb(int size, void *priv)
{
	struct data *buf = priv;

	if (size > 0) {
		buf->flag += size;
		buf = NULL;
	}
}

static void bluetooth_status_cb(int status)
{

	/* Check the USB status and do needed action if required */
	switch (status) {
	case 1:
		printf("USB device error");
		break;
	case 2:
		printf("USB device reset detected");
		break;
	case 3:
		printf("USB device connected");
		break;
	case 4:
		printf("USB device configured");
		/* Start reading */
		acl_read_cb(0, NULL);
		break;
	case 5:
		printf("USB device disconnected");
		/* Cancel any transfer */
		break;
	case 6:
		printf("USB device suspended");
		break;
	case 7:
		printf("USB device resumed");
		break;
	case 8:
		break;
	case 9:
	default:
		printf("USB unknown state");
		break;
	}
}
```

I put this `npd.c` in `npd/src` folder and then I use command `joern-parse -o cpg.bin ./npd/src`. However, I got the following errors:

```
Parsing code at: ./npd/src - language: `NEWC`
[+] Running language frontend
=======================================================================================================
Invoking CPG generator in a separate process. Note that the new process will consume additional memory.
If you are importing a large codebase (and/or running into memory issues), please try the following:
1) exit joern
2) invoke the frontend: /xxx/joern/joern-cli/c2cpg.sh -J-Xmx30208m ./npd/src --output cpg.bin
3) start joern, import the cpg: `importCpg("path/to/cpg")`
=======================================================================================================

[+] Applying default overlays
[INFO ] initializing 129 nodes from existing storage
[INFO ] installed GC monitors. will clear references if heap (after GC) is larger than 80%
[INFO ] Start of pass: io.joern.x2cpg.passes.base.FileCreationPass
[INFO ] Pass io.joern.x2cpg.passes.base.FileCreationPass completed in 10 ms (15% on mutations). 18 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.NamespaceCreator
[INFO ] Pass io.joern.x2cpg.passes.base.NamespaceCreator completed in 3 ms (19% on mutations). 4 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.TypeDeclStubCreator
[INFO ] Pass io.joern.x2cpg.passes.base.TypeDeclStubCreator completed in 1 ms (1% on mutations). 0 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.MethodStubCreator
[INFO ] Pass io.joern.x2cpg.passes.base.MethodStubCreator completed in 25 ms (34% on mutations). 70 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.ParameterIndexCompatPass
[INFO ] Pass io.joern.x2cpg.passes.base.ParameterIndexCompatPass completed in 0 ms (2% on mutations). 0 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.MethodDecoratorPass
[INFO ] Pass io.joern.x2cpg.passes.base.MethodDecoratorPass completed in 3 ms (34% on mutations). 57 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.AstLinkerPass
[INFO ] Pass io.joern.x2cpg.passes.base.AstLinkerPass completed in 8 ms (0% on mutations). 14 + 0 changes committed from 1 parts.
[INFO ] Start of enhancement: io.joern.x2cpg.passes.base.ContainsEdgePass
[INFO ] Enhancement io.joern.x2cpg.passes.base.ContainsEdgePass completed in 18 ms. 106  + 0 changes committed from 25 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.TypeRefPass
[INFO ] Pass io.joern.x2cpg.passes.base.TypeRefPass completed in 1 ms (3% on mutations). 7 + 0 changes committed from 1 parts.
[INFO ] Start of pass: io.joern.x2cpg.passes.base.TypeEvalPass
[ERROR] Pass io.joern.x2cpg.passes.base.TypeEvalPass failed
java.lang.IllegalArgumentException: requirement failed: size=0 and step=0, but both must be positive
        at scala.collection.Iterator$GroupedIterator.<init>(Iterator.scala:158) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.Iterator.grouped(Iterator.scala:345) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.Iterator.grouped$(Iterator.scala:344) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.AbstractIterator.grouped(Iterator.scala:1300) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.IterableOps.grouped(Iterable.scala:479) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.IterableOps.grouped$(Iterable.scala:478) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.AbstractIterable.grouped(Iterable.scala:933) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.x2cpg.passes.base.TypeEvalPass.generateParts(TypeEvalPass.scala:28) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at io.joern.x2cpg.passes.base.TypeEvalPass.generateParts(TypeEvalPass.scala:26) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at io.shiftleft.passes.NewStyleCpgPassBase.runWithBuilder(CpgPass.scala:129) ~[io.shiftleft.codepropertygraph_3-1.6.13.jar:1.6.13]
        at io.shiftleft.passes.ForkJoinParallelCpgPass.createApplySerializeAndStore(CpgPass.scala:72) ~[io.shiftleft.codepropertygraph_3-1.6.13.jar:1.6.13]
        at io.shiftleft.semanticcpg.layers.LayerCreator.runPass(LayerCreator.scala:48) ~[io.joern.semanticcpg-2.0.385.jar:2.0.385]
        at io.joern.x2cpg.layers.Base.create$$anonfun$1(Base.scala:37) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15) ~[org.scala-lang.scala3-library_3-3.4.2.jar:3.4.2]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10) ~[org.scala-lang.scala3-library_3-3.4.2.jar:3.4.2]
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1300) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.x2cpg.layers.Base.create(Base.scala:38) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at io.shiftleft.semanticcpg.layers.LayerCreator.run(LayerCreator.scala:32) ~[io.joern.semanticcpg-2.0.385.jar:2.0.385]
        at io.joern.x2cpg.X2Cpg$.applyDefaultOverlays$$anonfun$1(X2Cpg.scala:326) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15) ~[org.scala-lang.scala3-library_3-3.4.2.jar:3.4.2]
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10) ~[org.scala-lang.scala3-library_3-3.4.2.jar:3.4.2]
        at scala.collection.immutable.List.foreach(List.scala:333) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.x2cpg.X2Cpg$.applyDefaultOverlays(X2Cpg.scala:326) ~[io.joern.x2cpg-2.0.385.jar:2.0.385]
        at io.joern.joerncli.DefaultOverlays$.create(DefaultOverlays.scala:20) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at io.joern.joerncli.JoernParse$.applyDefaultOverlays$$anonfun$1(JoernParse.scala:164) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at scala.util.Try$.apply(Try.scala:210) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.joerncli.JoernParse$.applyDefaultOverlays(JoernParse.scala:169) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at io.joern.joerncli.JoernParse$.run$$anonfun$2$$anonfun$1$$anonfun$1(JoernParse.scala:91) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at scala.util.Success.flatMap(Try.scala:258) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.joerncli.JoernParse$.run$$anonfun$2$$anonfun$1(JoernParse.scala:92) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at scala.util.Success.flatMap(Try.scala:258) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.joerncli.JoernParse$.run$$anonfun$2(JoernParse.scala:92) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at scala.util.Success.flatMap(Try.scala:258) ~[org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.joerncli.JoernParse$.run(JoernParse.scala:92) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at io.joern.joerncli.JoernParse$.run$$anonfun$1(JoernParse.scala:78) ~[io.joern.joern-cli-2.0.385.jar:2.0.385]
        at scala.util.Success.flatMap(Try.scala:258) [org.scala-lang.scala-library-2.13.12.jar:?]
        at io.joern.joerncli.JoernParse$.run(JoernParse.scala:78) [io.joern.joern-cli-2.0.385.jar:2.0.385]
        at io.joern.joerncli.JoernParse$.main(JoernParse.scala:20) [io.joern.joern-cli-2.0.385.jar:2.0.385]
        at io.joern.joerncli.JoernParse.main(JoernParse.scala) [io.joern.joern-cli-2.0.385.jar:2.0.385]
[INFO ] Pass io.joern.x2cpg.passes.base.TypeEvalPass completed in 6 ms (0% on mutations). -1 + 0 changes committed from 0 parts.
java.lang.IllegalArgumentException: requirement failed: size=0 and step=0, but both must be positive
        at scala.collection.Iterator$GroupedIterator.<init>(Iterator.scala:158)
        at scala.collection.Iterator.grouped(Iterator.scala:345)
        at scala.collection.Iterator.grouped$(Iterator.scala:344)
        at scala.collection.AbstractIterator.grouped(Iterator.scala:1300)
        at scala.collection.IterableOps.grouped(Iterable.scala:479)
        at scala.collection.IterableOps.grouped$(Iterable.scala:478)
        at scala.collection.AbstractIterable.grouped(Iterable.scala:933)
        at io.joern.x2cpg.passes.base.TypeEvalPass.generateParts(TypeEvalPass.scala:28)
        at io.joern.x2cpg.passes.base.TypeEvalPass.generateParts(TypeEvalPass.scala:26)
        at io.shiftleft.passes.NewStyleCpgPassBase.runWithBuilder(CpgPass.scala:129)
        at io.shiftleft.passes.ForkJoinParallelCpgPass.createApplySerializeAndStore(CpgPass.scala:72)
        at io.shiftleft.semanticcpg.layers.LayerCreator.runPass(LayerCreator.scala:48)
        at io.joern.x2cpg.layers.Base.create$$anonfun$1(Base.scala:37)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1300)
        at io.joern.x2cpg.layers.Base.create(Base.scala:38)
        at io.shiftleft.semanticcpg.layers.LayerCreator.run(LayerCreator.scala:32)
        at io.joern.x2cpg.X2Cpg$.applyDefaultOverlays$$anonfun$1(X2Cpg.scala:326)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at io.joern.x2cpg.X2Cpg$.applyDefaultOverlays(X2Cpg.scala:326)
        at io.joern.joerncli.DefaultOverlays$.create(DefaultOverlays.scala:20)
        at io.joern.joerncli.JoernParse$.applyDefaultOverlays$$anonfun$1(JoernParse.scala:164)
        at scala.util.Try$.apply(Try.scala:210)
        at io.joern.joerncli.JoernParse$.applyDefaultOverlays(JoernParse.scala:169)
        at io.joern.joerncli.JoernParse$.run$$anonfun$2$$anonfun$1$$anonfun$1(JoernParse.scala:91)
        at scala.util.Success.flatMap(Try.scala:258)
        at io.joern.joerncli.JoernParse$.run$$anonfun$2$$anonfun$1(JoernParse.scala:92)
        at scala.util.Success.flatMap(Try.scala:258)
        at io.joern.joerncli.JoernParse$.run$$anonfun$2(JoernParse.scala:92)
        at scala.util.Success.flatMap(Try.scala:258)
        at io.joern.joerncli.JoernParse$.run(JoernParse.scala:92)
        at io.joern.joerncli.JoernParse$.run$$anonfun$1(JoernParse.scala:78)
        at scala.util.Success.flatMap(Try.scala:258)
        at io.joern.joerncli.JoernParse$.run(JoernParse.scala:78)
        at io.joern.joerncli.JoernParse$.main(JoernParse.scala:20)
        at io.joern.joerncli.JoernParse.main(JoernParse.scala)
```
The weird thing is that when I tested on another machine, the command successfully produced a cpg.bin. I use the same version of joern(2.0.385) , the same version of java(java19.0.2) and the same Linux distribution(ubuntu 22.04).  The machine where the error occurred used 2 AMD EPYC 7763, and the machine where the command was successfully executed used 4 Intel(R) Xeon(R) Gold 6130. I wonder if this is due to an environmental issue or a hardware issue.

**To Reproduce**
I'm not sure how to reproduce the issue because the same version of joern behaves differently on different machines.

**Expected behavior**
The command successfully generates a `cpg.bin` file.

**Desktop (please complete the following information):**
 - OS.  The machine where the `joern-parse` command was successfully executed:  `Linux 64Core 6.5.0-15-generic #15~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Jan 12 18:54:30 UTC 2 x86_64 x86_64 x86_64 GNU/Linux`
 - OS.  The machine where the `joern-parse` command failed:  `Linux k8s--master 5.15.0-48-generic #54-Ubuntu SMP Fri Aug 26 13:26:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux `
 - Joern Version.  both machines use joern 2.0.385
 - Java version.  both machines use java 19.0.2
 ```
$ java --version 
java 19.0.2 2023-01-17
Java(TM) SE Runtime Environment (build 19.0.2+7-44)
Java HotSpot(TM) 64-Bit Server VM (build 19.0.2+7-44, mixed mode, sharing)
$ javac --version
javac 19.0.2
```
**Additional Contexts**
hardware infomation for the machine where the `joern-parse` command failed:
```
CPU:  2 * AMD EPYC 7763 64-Core Processor
Memory: 1.0TiB
```
hardware infomation for the machine where the `joern-parse` command was successfully executed:
```
CPU: 4 * Intel(R) Xeon(R) Gold 6130 CPU
Memory: 187GiB
```

**Comment by wildoranges** [comment]
> To exclude multithreading issues (as we have seen on other recent bugreports) could you run it with `-J-XX:ActiveProcessorCount=1`?

with `-J-XX:ActiveProcessorCount=1` option, both two machines can successfully run the joern-parse command. Thanks.

**Comment by wildoranges** [comment]
> To exclude multithreading issues (as we have seen on other recent bugreports) could you run it with `-J-XX:ActiveProcessorCount=1`?

more info: It seems that joern-parse might not work properly with multiple CPUs. I was able to successfully generate cpg.bin when using `-J-XX:ActiveProcessorCount=64`. However, when using `-J-XX:ActiveProcessorCount=128`, the aforementioned error occurs(this machine has two AMD EPYC 7763 64-Core Processor).

**Comment by max-leuthaeuser** [comment]
@johannescoetzee @DavidBakerEffendi @bbrehm
So that worries me somewhat. We have quite some beefy machines here with tons of processors / RAM.
Could the implementation of the parallel passes (e.g., ForkJoinParallelCpgPass) be the issue here?
Or some limitation w.r.t. ODB with very high thread counts?

**Comment by wildoranges** [comment]
more info:  after my testing on my two machines, `-J-XX:ActiveProcessorCount=123` has no errors, but processor count > 123 will generate the aforementioned error. 123 seems to be a boundary.

**Comment by bbrehm** [comment]
Ok, the bug is [here](https://github.com/joernio/joern/blob/42b031ac44dd4d85336449041c8cabbb3c288cd3/joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/LinkingUtil.scala#L39).

If you run the pass with 7 nodes that require linking, and have 16 processors available, then you end up with a batch size of `7/16` which is zero. This throws in the grouped iterator thing.

There is a one-line quickfix -- just set the batchsize to 100 and be done with that nonsense, starting a thread ain't worth it for fewer items anyways. Or if you insist, just ensure that the batchsize is at least 1.

There is an actual fix as well: This code is really really bad. So I'll need to take a look how this got into the codebase in the first place, and talk to both author and reviewers and presumably write another "how not to shoot into your foot when multi-threading" primer.

I'm out with sick kid for today, so I'll look into it once I'm back. Feel free to merge a quickfix in the meantime.

**Comment by DavidBakerEffendi** [comment]
Possibly related? https://github.com/joernio/joern/issues/4596

According to this bug fix, it was introduced in [#4227](https://github.com/joernio/joern/pull/4227/files#diff-c83d92009d98201247cbe61e45ac8f8d2cc3a9cd40a20770d18da914dbc82a77R39), and unfortunately, @bbrehm, you were also a reviewer 😉 

It was a pretty big, complex, and drawn-out PR however, with lots of iterations, but mostly was customer impacting in that the original concurrency of this source finder was not scaling well.
